src/mono/mono/metadata/marshal.c

```diff
@@ -3555,133 +3574,130 @@ mono_marshal_get_native_wrapper (MonoMethod *method, gboolean check_exceptions,
-		mono_error_cleanup (emitted_error);
-
-		info = mono_wrapper_info_create (mb, WRAPPER_SUBTYPE_NONE);
-		info->d.managed_to_native.method = method;
-
-		csig = mono_metadata_signature_dup_full (get_method_image (method), sig);
-		csig->pinvoke = 0;
-		res = mono_mb_create_and_cache_full (cache, method, mb, csig,
-											 csig->param_count + 16, info, NULL);
-		mono_mb_free (mb);
-
-		return res;
-	}
-
 	/* internal calls: we simply push all arguments and call the method (no conversions) */
 	if (method->iflags & (METHOD_IMPL_ATTRIBUTE_INTERNAL_CALL | METHOD_IMPL_ATTRIBUTE_RUNTIME)) {
 		if (sig->hasthis)
 			csig = mono_metadata_signature_dup_add_this (get_method_image (method), sig, method->klass);
 		else
 			csig = mono_metadata_signature_dup_full (get_method_image (method), sig);
 
-		//printf ("%s\n", mono_method_full_name (method, 1));
-
 		/* hack - string constructors returns a value */
 		if (method->string_ctor)
 			csig->ret = string_type;
 
 		get_marshal_cb ()->emit_native_icall_wrapper (mb, method, csig, check_exceptions, aot, piinfo);
+	} else {
+		g_assert(pinvoke);
 
-		info = mono_wrapper_info_create (mb, WRAPPER_SUBTYPE_NONE);
-		info->d.managed_to_native.method = method;
+		csig = mono_metadata_signature_dup_full (get_method_image (method), sig);
+		mono_marshal_set_callconv_from_modopt (method, csig, FALSE);
 
-		csig = mono_metadata_signature_dup_full (get_method_image (method), csig);
-		csig->pinvoke = 0;
-		res = mono_mb_create_and_cache_full (cache, method, mb, csig, csig->param_count + 16,
-											 info, NULL);
+		mspecs = g_new0 (MonoMarshalSpec*, sig->param_count + 1);
+		mono_method_get_marshal_info (method, mspecs);
 
-		mono_mb_free (mb);
-		return res;
-	}
+		if (mono_class_try_get_suppress_gc_transition_attribute_class ()) {
+			MonoCustomAttrInfo *cinfo;
+			ERROR_DECL (error);
 
-	g_assert (pinvoke);
+			cinfo = mono_custom_attrs_from_method_checked (method, error);
+			mono_error_assert_ok (error);
+			gboolean found = FALSE;
+			if (cinfo) {
+				for (int i = 0; i < cinfo->num_attrs; ++i) {
+					MonoClass *ctor_class = cinfo->attrs [i].ctor->klass;
+					if (ctor_class == mono_class_try_get_suppress_gc_transition_attribute_class ()) {
+						found = TRUE;
+						break;
+					}
+				}
+			}
+			if (found)
+				skip_gc_trans = TRUE;
+			if (cinfo && !cinfo->cached)
+				mono_custom_attrs_free (cinfo);
+		}
 
-	csig = mono_metadata_signature_dup_full (get_method_image (method), sig);
-	mono_marshal_set_callconv_from_modopt (method, csig, FALSE);
+		if (csig->call_convention == MONO_CALL_DEFAULT) {
+			/* If the calling convention has not been set, check the UnmanagedCallConv attribute */
+			mono_marshal_set_callconv_from_unmanaged_callconv_attribute (method, csig, &skip_gc_trans);
+		}
 
-	mspecs = g_new0 (MonoMarshalSpec*, sig->param_count + 1);
-	mono_method_get_marshal_info (method, mspecs);
+		if (mono_method_signature_has_ext_callconv (csig, MONO_EXT_CALLCONV_SWIFTCALL)) {
```

Comment:
@kotlarmilos @vargaz Do we need to prevent these m2n swiftcall wrappers from being inlined?  I think there's logic in the mini-arm64 and mini-amd64 backends that depends on the `pinvoke` bit of the signature.

Reply:
Is the inlining performed before or after the lowering process? Do the method prolog and epilog get inlined as well?

ArgSwiftError is handled differently in the method prolog, which could be affected when it is inlined. Also, there is a potential problem related to context registers clobbering when wrappers are inlined.

Reply:
These wrappers are not inlined right now.